common/reactive/src/main/java/io/helidon/common/reactive/EmittingPublisher.java

```diff
@@ -0,0 +1,334 @@
+                if (state.get() == State.CANCELLED) {
+                    return;
+                }
+                if (n < 1) {
+                    fail(new IllegalArgumentException("Rule §3.9 violated: non-positive request amount is forbidden"));
+                    return;
+                }
+                requested.updateAndGet(r -> Long.MAX_VALUE - r > n ? n + r : Long.MAX_VALUE);
+                state.compareAndSet(State.NOT_REQUESTED_YET, State.READY_TO_EMIT);
+                requestCallback.accept(n, requested.get());
+            }
+
+            @Override
+            public void cancel() {
+                if (state.compareAndSet(State.NOT_REQUESTED_YET, State.CANCELLED)
+                        || state.compareAndSet(State.READY_TO_EMIT, State.CANCELLED)) {
+                    cancelCallback.run();
+                    EmittingPublisher.this.subscriber = null;
+                }
+            }
+
+        });
+        deferredComplete.complete(null);
+    }
+
+    /**
+     * Properly fail the stream, set publisher to cancelled state and send {@code onError} signal downstream.
+     * Signal {@code onError} is sent only once, any other call to this method is no-op.
+     *
+     * @param throwable Sent as {@code onError} signal
+     */
+    public void fail(Throwable throwable) {
+        if (deferredComplete.isDone()) {
+            signalOnError(throwable);
+        } else {
+            deferredComplete.thenRun(() -> signalOnError(throwable));
+        }
+    }
+
+    /**
+     * Properly complete the stream, set publisher to completed state and send {@code onComplete} signal downstream.
+     * Signal {@code onComplete} is sent only once, any other call to this method is no-op.
+     */
+    public void complete() {
+        deferredComplete.thenRun(this::signalOnComplete);
+    }
+
+    private void signalOnError(Throwable throwable) {
+        if (state.compareAndSet(State.NOT_REQUESTED_YET, State.FAILED)
+                || state.compareAndSet(State.READY_TO_EMIT, State.FAILED)) {
+            for (;;) {
+                try {
+                    if (emitting.getAndSet(true)) {
+                        continue;
+                    }
+                    this.subscriber.onError(throwable);
+                    return;
+                } catch (Throwable t) {
+                    throw new IllegalStateException("On error threw an exception!", t);
+                } finally {
+                    emitting.set(false);
+                }
+            }
+        }
+    }
+
+    private void signalOnComplete() {
+        if (state.compareAndSet(State.NOT_REQUESTED_YET, State.COMPLETED)
+                || state.compareAndSet(State.READY_TO_EMIT, State.COMPLETED)) {
+            for (;;) {
+                try {
+                    if (emitting.getAndSet(true)) {
+                        continue;
+                    }
+                    this.subscriber.onComplete();
+                    return;
```

Comment:
The subscriber reference is not released when completed.
One way of fixing it would be to add `EmittingPublisher.this.subscriber = null;` here

Reply:
Uf thx for that, there was also possible NPE if emit started right before `READY_TO_EMIT` -> `CANCELLED` state transition, but invoked `onNext` after reference release